crates/sui-framework/sources/balance.move

```diff
@@ -27,15 +31,39 @@ module sui::balance {
         self.value
     }
 
+    /// Get the `Supply` value.
+    public fun supply_value<T>(supply: &Supply<T>): u64 {
+        supply.value
+    }
+
+    /// Create a new supply for type T.
+    public fun create_supply<T: drop>(_witness: T): Supply<T> {
+        Supply { value: 0 }
+    }
+
+    /// Increase supply by `value` and create a new `Balance<T>` with this value.
+    public fun increase_supply<T>(supply: &mut Supply<T>, value: u64): Balance<T> {
+        supply.value = supply.value + value;
+        Balance { value }
+    }
+
+    /// Burn a Balance<T> and decrease Supply<T>.
+    public fun decrease_supply<T>(supply: &mut Supply<T>, balance: Balance<T>): u64 {
+        let Balance { value } = balance;
+        supply.value = supply.value - value;
```

Comment:
Let's assert this will not underflow

Reply:
It is possible if someone misuses `Supply` (e.g., creates two `Supply`'s)